dive_core/common/emulate_pm4.cpp

```diff
@@ -929,6 +929,114 @@ bool EmulatePM4::AdvanceCb(const IMemoryManager &mem_manager,
                 return false;
         }
     }
+    else if ((type == Pm4Type::kType7) && (type7_header.opcode == CP_START_BIN))
+    {
+
+        PM4_CP_START_BIN packet;
+        DIVE_VERIFY(mem_manager.CopyMemory(&packet,
+                                           emu_state_ptr->m_submit_index,
+                                           emu_state_ptr->GetCurIb()->m_cur_va,
+                                           sizeof(PM4_CP_START_BIN)));
+
+        // The CP_START_BIN & CP_END_BIN are pm4s only availabe at a650+
+        // here is the layout:
+        //  |CP_START_BIN|
+        //  |Prefix_block_0|
+        //  |Prefix_block_1|
+        //  ...
+        //  |Prefix_block_n|
+        //  |Common_block|
+        //  |CP_END_BIN|
+        // All the prefix blocks are in a separate IB
+        // Prefix_block contains setup for each tile, like viewport/scissors with fixed block size
+        // Common_block contains drawcalls and resolve
+
+        // only add the 1st prefix block here
+        // after this ib, we jump back to the Common_block
+        // once we advance to the CP_END_BIN, we add the rest ibs
+        // it is done this way since we don't know the size of the Common_block at this point
+        if (!QueueIB(packet.PREFIX_ADDR,
+                     packet.PREFIX_DWORDS,
+                     false,
+                     IbType::kBinPrefix,
+                     emu_state_ptr))
+        {
+            return false;
+        }
+
+        m_cp_start_bin_address = emu_state_ptr->GetCurIb()->m_cur_va +
+                                 GetPacketSize(header) * sizeof(uint32_t);
+        m_prefix_start_address = packet.PREFIX_ADDR;
+        m_prefix_block_dword_size = packet.PREFIX_DWORDS;
+        m_prefix_block_count = packet.BIN_COUNT;
+
+        AdvancePacket(emu_state_ptr, header);
+        if (!AdvanceToQueuedIB(mem_manager, emu_state_ptr, callbacks))
+            return false;
+    }
+    else if ((type == Pm4Type::kType7) && (type7_header.opcode == CP_END_BIN))
+    {
+        uint32_t common_block_dword_size = static_cast<uint32_t>(
+        (emu_state_ptr->GetCurIb()->m_cur_va - m_cp_start_bin_address) / sizeof(uint32_t));
+        // starts from the 2nd one since the 1st one has been added in CP_START_BIN
+        for (uint32_t bin = 1; bin < m_prefix_block_count; ++bin)
+        {
+            if (!QueueIB(m_prefix_start_address +
+                         bin * m_prefix_block_dword_size * sizeof(uint32_t),
+                         m_prefix_block_dword_size,
+                         false,
+                         IbType::kBinPrefix,
+                         emu_state_ptr))
+            {
+                return false;
+            }
+
+            if (!QueueIB(m_cp_start_bin_address,
```

Comment:
The way it's written here, you have a prefix IB followed by a common IB, but only for all tiles except the first one.
The first one, which is covered by the CP_START_BIN, only covers a kBinPrefix but no kBinCommon. So that means there's no kBinCommon called for the first tile? Am I missing something?

Reply:
since i don't know the size of the common block, i need to parse it. Note that all prefix blocks are in 1 IB, but I intentionally break it into multiple ones so that it is easier to parse. when i reach CP_START_BIN, i only add the 1st prefix block as IB, and start to parse the 1st prefix block, once it is done, the parser jump back to the common block since it is right after the IB of all prefix blocks, and now i am at the start of the common block. once i reach CP_END_BIN, i am reaching the end of common block, and now i have the start and size of the common block. at this point, i have already added all the packets for the 1st bin, that is why i skipped the 1st tile.

Reply:
Ahhhh ok. The common block is implicitly parsed as the next packets right after that 1st common block. Makes sense. I missed that part!